--- /usr/bin/pt-stalk	2012-06-12 09:04:22.000000000 -0700
+++ ./pt-stalk-raingauge	2012-09-05 11:14:58.948416483 -0700
@@ -783,20 +783,21 @@
       (echo $ts; df -h) >> "$d/$p-df" &
 
       (echo $ts; netstat -antp) >> "$d/$p-netstat"   &
       (echo $ts; netstat -s)    >> "$d/$p-netstat_s" &
 
       (echo $ts; $CMD_MYSQL $EXT_ARGV -e "SHOW FULL PROCESSLIST\G") \
          >> "$d/$p-processlist" &
 
       if [ "$have_lock_waits_table" ]; then
          (echo $ts; lock_waits) >>"$d/$p-lock-waits" &
+         (echo $ts; locks) >> "$d/$p-locks" &
       fi
    done
    log "Loop end: $(date +'TS %s.%N %F %T')"
 
    if [ "$have_oprofile" ]; then
       $CMD_OPCONTROL --stop
       $CMD_OPCONTROL --dump
 
       local oprofiled_pid=$(_pidof oprofiled | awk '{print $1; exit;}')
       if [ "$oprofiled_pid" ]; then
@@ -881,33 +882,38 @@
       SUBSTRING(p.host, INSTR(p.host, ':') +1) AS blocking_port,
       IF(p.command = \"Sleep\", p.time, 0) AS idle_in_trx,
       b.trx_query AS blocking_query
    FROM INFORMATION_SCHEMA.INNODB_LOCK_WAITS AS w
    INNER JOIN INFORMATION_SCHEMA.INNODB_TRX AS b ON b.trx_id = w.blocking_trx_id
    INNER JOIN INFORMATION_SCHEMA.INNODB_TRX AS r ON r.trx_id = w.requesting_trx_id
    INNER JOIN INFORMATION_SCHEMA.INNODB_LOCKS AS l ON w.requested_lock_id = l.lock_id
    LEFT JOIN INFORMATION_SCHEMA.PROCESSLIST AS p ON p.id = b.trx_mysql_thread_id
    ORDER BY wait_time DESC\G"
    $CMD_MYSQL $EXT_ARGV -e "$sql2"
-} 
+}
+
+locks() {
+   local sql="SELECT * FROM INFORMATION_SCHEMA.INNODB_LOCKS"
+   $CMD_MYSQL $EXT_ARGV -e "$sql"
+}
 
 # ###########################################################################
 # End collect package
 # ###########################################################################
 
 # ###########################################################################
 # Global variables
 # ###########################################################################
 TRIGGER_FUNCTION=""
 RAN_WITH=""
 EXIT_REASON=""
-TOOL="pt-stalk"
+TOOL="pt-stalk-raingauge"
 OKTORUN=1
 ITER=1
 
 # ###########################################################################
 # Subroutines
 # ###########################################################################
 
 grep_processlist() {
    local file="$1"
    local col="$2"
@@ -996,21 +1002,21 @@
       EXIT_REASON="no more iterations"
       return 1  # stop running
    fi
 
    return 0  # continue running
 }
 
 sleep_ok() {
    local seconds="$1"
    local msg="${2:-""}"
-   if oktorun; then
+   if oktorun || [[ "$OPT_EXEC_AFTER_SLEEP" ]]; then
       if [ -n "$msg" ]; then
          log "$msg"
       fi
       sleep $seconds
    fi
 }
 
 purge_samples() {
    local dir="$1"
    local retention_time="$2"
@@ -1121,29 +1127,36 @@
                warn "Collect canceled because there will not be enough disk space after collecting another $margin MB"
             fi
          fi
 
          # ##################################################################
          # Done collecting.
          # ##################################################################
          ITER=$((ITER + 1))
          cycles_true=0
          sleep_ok "$OPT_SLEEP" "Sleeping $OPT_SLEEP seconds after collect"
+
+         if [ "$OPT_EXEC_AFTER_SLEEP" ]; then
+	   log "Executing script after sleep: $OPT_EXEC_AFTER_SLEEP"
+	   $OPT_EXEC_AFTER_SLEEP
+	 fi
+
+	 # Purge old collect file between checks.
+	 if [ -d "$OPT_DEST" ]; then
+	     purge_samples "$OPT_DEST" "$OPT_RETENTION_TIME"
+	 fi
+
       else
          # Trigger/check/value is ok, sleep until next check.
          sleep_ok "$OPT_INTERVAL"
       fi
 
-      # Purge old collect file between checks.
-      if [ -d "$OPT_DEST" ]; then
-         purge_samples "$OPT_DEST" "$OPT_RETENTION_TIME"
-      fi
    done
 }
 
 # ###########################################################################
 # Main program loop, called below if tool is ran from the command line.
 # ###########################################################################
 
 main() {
    trap sigtrap SIGHUP SIGINT SIGTERM
 
@@ -1186,21 +1199,21 @@
 
    # Verify and set TRIGGER_FUNCTION based on --function.
    if ! set_trg_func "$OPT_FUNCTION"; then
       option_error "Invalid --function value: $OPT_FUNCTION"
    fi
 
    if [ -z "$OPT_STALK" -a "$OPT_COLLECT" ]; then
       # Not stalking; do immediate collect once.
       OPT_ITERATIONS=1
       OPT_CYCLES=0
-      OPT_SLEEP=0
+      #OPT_SLEEP=0
       OPT_INTERVAL=0
    fi
 
    usage_or_errors "$0"
    po_status=$?
    rm_tmpdir
    if [ $po_status -ne 0 ]; then
       [ $OPT_ERRS -gt 0 ] && exit 1
       exit 0
    fi
@@ -1477,20 +1490,33 @@
 type: int; default: 5
 
 Don't collect data if the disk has less than this percent free space.
 This prevents the tool from filling up the disk with diagnostic data.
 
 This option works similarly to L<"--disk-bytes-free"> but specifies a
 percentage margin of safety instead of a bytes margin of safety.
 The tool honors both options, and will not collect any data unless both
 margins are satisfied.
 
+=item --exec-after-collect
+
+type: string
+
+Run this script immediately after the collection has been started.
+
+
+=item --exec-after-sleep
+
+type: string
+
+Run this script after collection has been completed, and pt-stalk has waited for L<"--sleep"> seconds.
+
 =item --function
 
 type: string; default: status
 
 Specifies what to watch for a diagnostic trigger.  The default value watches
 SHOW GLOBAL STATUS, but you can also watch SHOW PROCESSLIST or supply a plugin
 file with your own custom code.  This function supplies the value of
 L<"--variable">, which is then compared against L<"--threshold"> to see if the
 trigger condition is met.  Additional options may be required as well; see
 below. Possible values:
